src/client/metadataTransfer.ts

```diff
@@ -269,3 +269,16 @@ export abstract class MetadataTransfer<Status extends MetadataRequestStatus, Res
   protected abstract pre(): Promise<AsyncResult>;
   protected abstract post(result: Status): Promise<Result>;
 }
+
+/* prevent requests on apiVersions higher than the org supports */
+const getConnectionNoHigherThanOrgAllows = async (conn: Connection, requestedVersion: string): Promise<Connection> => {
+  // uses a TTL cache, so mostly won't hit the server
+  const maxApiVersion = await conn.retrieveMaxApiVersion();
+  if (requestedVersion && parseInt(requestedVersion, 10) > parseInt(maxApiVersion, 10)) {
```

Comment:
Would it be worthwhile to validate the requested version here?  It seems the caller could send an invalid version string here and get by without altering the max version allowed on the connection.

Reply:
the ComponentSet allows consumers to set its apiVersion and sourceApiVersion properties.  If there's a place to put validation, it'd be there.  This is internal non-exported function, only called from this file, which is extended by Deploy and Retrieve.

Are you wanting to put getters/setters so we can do validation on those props for ComponentSet?